### ISSUE TITLE
Compiler hits infinite loop on code after e03d254bbd54

### ISSUE THREAD

**Issue by dyung** [llvm/llvm-project#61224]
We have an internal test case which recently has started hanging when building on our machines and I finally got a chance to investigate it. When I looked into it, the compiler seems to be getting into an infinite loop, and with some help from @gregbedwell  I was able to reduce it to the following code:
```c++
double a(double c, double d) { return c >= d ? c : d; }
struct e {
  double f;
  void operator*(double);
} b;
struct {
  e g;
  double h();
  double i() {
    double j, k = h(), l = g.f >= 0 ? g.f : 0;
    j = k >= l ? k : l;
    return j;
  }
} m;
void n() {
  double o = m.i(), p = a(0, o);
  b *p;
}
```
If compiled with optimizations enabled and fast-math (`-O2 -fast-math`), the compilation hangs until you kill it or your system runs out of memory. I was able to bisect this issue back to commit e03d254bbd54.

**Comment by RKSimon** [comment]
CC @alexey-bataev

**Comment by alexey-bataev** [comment]
Fixed in c411965820eb803dd7eac39f80357cad663b7ba0
